### PR TITLE
net: tcp: Restructure the close FIN sequence

### PR DESCRIPTION
**Pull request by ssharks**
When the application closes the connection, it enters the FIN1 state.
Restructure this sequence to RST on data received after the application
has closed the socket.
Next to that deal with packets that contain both data and a FIN flag.

### CODE REVIEW

File: tests/net/tcp/src/main.c [hunk 1]
```diff
@@ -1118,6 +1128,269 @@ ZTEST(net_tcp, test_client_fin_wait_2_ipv4)
...
+
+	ret = net_context_connect(ctx, (struct sockaddr *)&peer_addr_s,
+				  sizeof(struct sockaddr_in),
+				  NULL,
+				  K_MSEC(100), NULL);
+	if (ret < 0) {
+		zassert_true(false, "Failed to connect to peer");
+	}
+
+	ret = net_context_recv(ctx, test_tcp_recv_cb, K_MSEC(200), NULL);
```
Comment (ssharks) [inline tests/net/tcp/src/main.c:1258]: @rlubos: While rebasing I found that the latest commit of this PR breaks. This context_recv never completes. It is the only place in these tests where the recv is called from a client context. Might it have something to do with the recently merged PR: https://github.com/zephyrproject-rtos/zephyr/pull/61336

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+					 */
+					new_len++;
 				}
 			}
 
 			/*
-			 * Only if our FIN flag has been acknowledged and all the data has been
-			 * received, it can go to the TIME_WAIT state
+			 * FIN1:
+			 * Acknowledge path and sequence path are independent, threat them that way
```
Comment (AndreyDodonov-EH) [inline subsys/net/ip/tcp.c:2659]: Typo: `treat`

Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

**Comment by AndreyDodonov-EH** [comment]
Just linking the problem which was observed by me in Zephyr and which had been documented in RFC long time ago for other clients.
I think it should be fixed by this PR.
https://www.rfc-editor.org/rfc/rfc2525#page-51
(in turn mentioned in https://www.rfc-editor.org/rfc/rfc9293#name-half-closed-connections)

In short, failure to send RST leads to connections staying open and eventually we run out of connections.
Currently (in 3.4.0) Zephyr keeps ack-ing data, mimicking normal Half-Closed connection, so it's eventually closed and thus not leaked. But data are not forwarded to the application.

File: tests/net/tcp/src/main.c (hunk 1 above)
Comment (rlubos) [reply tests/net/tcp/src/main.c:1258]: I've debuged this a bit and it seems like a race condition in the test. The DATA packet is received by the TCP stack **before** the test actually installs recv callback with `net_context_recv()`, so it gets dropped:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L982

The change from https://github.com/zephyrproject-rtos/zephyr/pull/61336 merely revealed that the `net_context_recv()` times out, as before that the timeout result was not returned correctly.

Comment (rlubos) [reply tests/net/tcp/src/main.c:1258]: Calling `net_context_recv()` before connect, with K_NO_WAIT timeout (just to install the recv cb) seems to help. The semaphore takes care of checking whether the data actually arrived.

**Comment by ssharks** [comment]
Apparently I'm no friends with the testing framework. To me it is totally unclear how the scheduling between the test, handler and TCP stack takes place with breaks, returns and goto's. I find in very hard to make reliable tests, 
It looks like I'm not the only one as the test_client_closing_ipv6 is not going through the closing state for example.

I'll stop here. @AndreyDodonov-EH  could you give it a spin on the real HW?

File: tests/net/tcp/src/main.c (hunk 1 above)
Comment (ssharks) [reply tests/net/tcp/src/main.c]: Thanks for your input. Finally I removed the `net_context_recv()` altogether as is not required to test the network stack.

**Comment by ssharks** [comment]
No luck with upstream merges

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
 
-				/* Acknowledge any received data */
-				if (len > 0) {
-					tcp_out(conn, ACK);
+			/*
+			 * FIN2:
+			 * Only FIN is relevant in this state, as FIN has already acknowledged
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2669]:
```suggestion
			 * Only FIN is relevant in this state, as FIN was already acknowledged
```

File: subsys/net/ip/tcp.c [hunk 4]
```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+			    (conn->state == TCP_CLOSING)) {
 				/* Cases:
 				 * - Data already received earlier: len > 0 , new_len <= 0
 				 * - Partially new data len > 0, new_len > 0
 				 * - Out of order data len > 0, new_len > 0, ignore the data in
 				 *   between
 				 */
 				if (new_len > 0) {
-					conn_ack(conn, + new_len);
+					/* This should not happen, since we do not implement half
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2707]: Is `This should not happen` correct here? Sure, we do not support half-closed streams, but `This should not happen` indicates that the peer did something against the standard.

File: subsys/net/ip/tcp.c [hunk 5]
```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
 				verdict = NET_OK;
 			}
-		}
-		break;
-	case TCP_FIN_WAIT_2:
-		/* Acknowledge but drop any data, but subtract any duplicate data */
-		if (th) {
-			if (len > 0) {
-				int32_t new_len = len - net_tcp_seq_cmp(conn->ack, th_seq(th));
+			if ((next == TCP_FIN_WAIT_2) ||
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2697]: What about `TCP_FIN_WAIT_1` state? Shouldn't we reject new data in that state as well? RFC 9293 [3.6.1](https://www.rfc-editor.org/rfc/rfc9293#section-3.6.1):
> A host MAY implement a "half-duplex" TCP close sequence, so that an application that has called CLOSE cannot continue to read data from the connection (MAY-1).

I understand that this means we should be sending RST as soon as we close the connection on our side (no need to wait for ACK for our FIN here?)

**Review by rlubos — COMMENTED** [review]
What is the rationale for creating a "combo" state handler for FIN1/FIN2/CLOSING states? To be honest, for me it's a bit hard to follow with all of the ifs within it. Was there a lot of code duplication by having separate state handlers? Could it perhaps be approach some other way, by for example creating helper functions for the overlapping functionality instead?

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c:2697]: @rlubos
I interpreted it so that we wait for ACK for our FIN normally, but should the new data be arriving from the host, we add RST (`...while received data is still pending in the TCP connection, or if new data is received after CLOSE is called, its TCP implementation SHOULD send a RST to show that data was lost (SHLD-3)...`)


But according to RFC 9293 [3.5.3 Reset Processing](https://www.rfc-editor.org/rfc/rfc9293#name-reset-processing)](https://www.rfc-editor.org/rfc/rfc9293#section-3.5.3) receiver goes to the CLOSED state, so we do not have to ACK new data, instead we can keep sending RST.

This way at least there is no connection leak, but we better be sure that our RST has arrived, otherwise we might have a leak at the host.

But better me graceful and sync properly, just add RST to show we do not process incoming data anymore.

**Comment by ssharks** [comment]
> What is the rationale for creating a "combo" state handler for FIN1/FIN2/CLOSING states? To be honest, for me it's a bit hard to follow with all of the ifs within it. Was there a lot of code duplication by having separate state handlers? Could it perhaps be approach some other way, by for example creating helper functions for the overlapping functionality instead?

Initially I thought it would be better this way, as the pattern is pretty similar throughout all 3 states, but adding it all together it becomes pretty messy indeed.
The risk in seperating is that you do not cover one case at one location.
I'll rewrite this to separate cases using helper functions.

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2697]: I was wondering about this behavior as well, on one hand you want to reset directly, but in FIN1, there is an inflight FIN flag. If we reset right away, we will "overwrite" the inflight FIN flag. The zephyr stack only accepts in sequence RST messages for instance.
This was my rationale for not resetting in FIN1 when presented with new data, but instead acking the last accepted data.

**Comment by AndreyDodonov-EH** [comment]
> Apparently I'm no friends with the testing framework. To me it is totally unclear how the scheduling between the test, handler and TCP stack takes place with breaks, returns and goto's. I find in very hard to make reliable tests, It looks like I'm not the only one, as the test_client_closing_ipv6 is not going through the closing state for example.
> 
> I'll stop here. @AndreyDodonov-EH could you give it a spin on the real HW?

Just cherry-picking these changes did not go well.
But I don't have time to update properly at the moment - too much to merge.

In case of cherry-picking first I get
`<err> net_tcp: conn: 0x2003a70c, new bytes 31 during FIN states sending reset
`

Which is expected, I guess, may be shouldn't even be an error.
But then buffers seem to leak.
`
[00:00:27.110,260] <err> net_pkt: Data buffer (1502) allocation failed.
[00:00:27.110,290] <err> sl_wfx_host: Failed to get net buffer
[00:00:57.830,413] <err> net_pkt: Data buffer (1502) allocation failed.
[00:00:57.830,444] <err> sl_wfx_host: Failed to get net buffer
`
And then it stays like that.
I'm not sure if it's because of this change or because I just cherry-picked, without properly updating.

**Comment by ssharks** [comment]
> > Apparently I'm no friends with the testing framework. To me it is totally unclear how the scheduling between the test, handler and TCP stack takes place with breaks, returns and goto's. I find in very hard to make reliable tests, It looks like I'm not the only one, as the test_client_closing_ipv6 is not going through the closing state for example.
> > I'll stop here. @AndreyDodonov-EH could you give it a spin on the real HW?
> 
> Just cherry-picking these changes did not go well. But I don't have time to update properly at the moment - too much to merge.
> 
> In case of cherry-picking first I get `<err> net_tcp: conn: 0x2003a70c, new bytes 31 during FIN states sending reset `
> 
> Which is expected, I guess, may be shouldn't even be an error. But then buffers seem to leak. `[00:00:27.110,260] <err> net_pkt: Data buffer (1502) allocation failed. [00:00:27.110,290] <err> sl_wfx_host: Failed to get net buffer [00:00:57.830,413] <err> net_pkt: Data buffer (1502) allocation failed. [00:00:57.830,444] <err> sl_wfx_host: Failed to get net buffer` And then it stays like that. I'm not sure if it's because of this change or because I just cherry-picked, without properly updating.

Thanks for your feedback I'll at least remove the error message. An look at possible buffer leaks

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c:2697]: @rlubos @ssharks 

Here https://www.rfc-editor.org/rfc/rfc9293#section-3.5.2 it says - 
The side of a connection issuing a reset should enter the TIME-WAIT state, as this generally helps to reduce the load on busy servers for reasons described in [[70](https://www.rfc-editor.org/rfc/rfc9293#FTY99)].

So I'd say once we get data we can not process, instead of ACK-ing them, we issue RST and go to TIME-WAIT (in which we should probably keep sending RST???)

File: tests/net/tcp/src/main.c [hunk 6]
```diff
@@ -1118,6 +1128,275 @@ ZTEST(net_tcp, test_client_fin_wait_2_ipv4)
...
+	if (ret < 0) {
+		zassert_true(false, "Failed to connect to peer");
+	}
+
+	/*
+	 * Disable the receive statement, as I cannot get it to work and it is not
+	 * required to get the test functional, the receive functionality is covered
+	 * by other tests.
+	 *
+	 * ret = net_context_recv(ctx, test_tcp_recv_cb, K_MSEC(200), NULL);
```
[the remark this answers is not in the document]
Comment (ssharks) [reply tests/net/tcp/src/main.c]: It indeed looks like the same issue. For this PR it is not blocking, but it looks to me like a significant shortcoming. At least worth an open issue. How to proceed here?
I processed your comment, modified the comment and removed the commented out code.

File: subsys/net/ip/tcp.c [hunk 7]
```diff
@@ -2639,117 +2639,138 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+					new_len++;
 				}
 			}
 
 			/*
-			 * Only if our FIN flag has been acknowledged and all the data has been
-			 * received, it can go to the TIME_WAIT state
+			 * FIN1:
+			 * Acknowledge path and sequence path are independent, treat them that way
+			 * The for combinations and their destination states
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Please check if you can find you more in the updated comment.

File: subsys/net/ip/tcp.c (hunk 3 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 4 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: True, only in the CLOSING state it should not happen, in FIN1 and FIN2 it is allowed. Modified the comment.

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: I modified the code to also reset from FIN1. I can see the point of going to TIME_WAIT after a reset. I think this is out of scope for this PR as it affects all the places where a RST packet is generated and I think also the sequence number checking for incoming RST packet should be broadened in my opinion to work well.
Is it worth creating a new issue for this? @AndreyDodonov-EH maybe you can submit a PR for this.

**Comment by ssharks** [comment]
> > > Apparently I'm no friends with the testing framework. To me it is totally unclear how the scheduling between the test, handler and TCP stack takes place with breaks, returns and goto's. I find in very hard to make reliable tests, It looks like I'm not the only one, as the test_client_closing_ipv6 is not going through the closing state for example.
> > > I'll stop here. @AndreyDodonov-EH could you give it a spin on the real HW?
> > 
> > 
> > Just cherry-picking these changes did not go well. But I don't have time to update properly at the moment - too much to merge.
> > In case of cherry-picking first I get `<err> net_tcp: conn: 0x2003a70c, new bytes 31 during FIN states sending reset `
> > Which is expected, I guess, may be shouldn't even be an error. But then buffers seem to leak. `[00:00:27.110,260] <err> net_pkt: Data buffer (1502) allocation failed. [00:00:27.110,290] <err> sl_wfx_host: Failed to get net buffer [00:00:57.830,413] <err> net_pkt: Data buffer (1502) allocation failed. [00:00:57.830,444] <err> sl_wfx_host: Failed to get net buffer` And then it stays like that. I'm not sure if it's because of this change or because I just cherry-picked, without properly updating.
> 
> Thanks for your feedback I'll at least remove the error message. An look at possible buffer leaks

I have had a quick look, but do not see how it could leak buffers quickly. The code in in that respect little different from what it used to be I think.

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: After thinking on it a bit I implemented at least the transition to TIME-WAIT from the FIN states when a resettable packet arrives. Fixing this revealed some problems with the timer, but I think I fixed it correct now.
The further I dig, the more surprised I'm that it worked reasonable in the past.

File: subsys/net/ip/tcp.c [hunk 8]
```diff
@@ -2639,124 +2666,203 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-					fl, conn->seq, conn->ack);
-				if (len > 0) {
+				} else {
+					/* Already acknowledged, we can go further */
+					NET_DBG("FIN received, going to CLOSING as no ACK has been "
+						"received");
 					tcp_send_timer_cancel(conn);
-					/* Send out a duplicate ACK, with the pending FIN flag */
-					tcp_out(conn, FIN | ACK);
+					tcp_out_ext(conn, FIN | ACK, NULL, conn->seq - 1);
```
Comment (ssharks) [inline subsys/net/ip/tcp.c:2719]: @rlubos: Do we need a send timer cancel here to avoid a buffer leak?

Comment (tbursztyka) [reply subsys/net/ip/tcp.c:2719]: isn't this done already just one line above? Or perhaps your comment is misplaced and should be about line 2734?

File: subsys/net/ip/tcp.c [hunk 9]
```diff
@@ -1960,6 +1960,33 @@ static bool tcp_validate_seq(struct tcp *conn, struct tcphdr *hdr)
 		(net_tcp_seq_cmp(th_seq(hdr), conn->ack + conn->recv_win) < 0);
 }
 
+static int32_t tcp_compute_new_length(struct tcp *conn, struct tcphdr *hdr, size_t len)
+{
+	int32_t new_len = 0;
+
+	if (len > 0) {
+		/* Cases:
+		 * - Data already received earlier: len > 0 , new_len <= 0
```
Comment (tbursztyka) [inline subsys/net/ip/tcp.c:1969]: Could rewrite the comment: all len > 0 are obvious here since we are in a if statement that needs it.

File: tests/net/tcp/src/main.c [hunk 10]
```diff
@@ -126,6 +126,7 @@ static void handle_server_test(sa_family_t af, struct tcphdr *th);
 static void handle_syn_resend(void);
```
Comment (tbursztyka) [inline tests/net/tcp/src/main.c:126]: add in the commit message that you also fixed the th_seq endian conversion.

File: subsys/net/ip/tcp.c (hunk 9 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: tests/net/tcp/src/main.c (hunk 10 above)
Comment (ssharks) [reply tests/net/tcp/src/main.c:126]: Done

File: subsys/net/ip/tcp.c [hunk 11]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+
+				verdict = NET_OK;
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
+				if (next == TCP_TIME_WAIT) {
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2823]: Maybe introduce `fin_acked` helper variable in this state as well, as suggested above?

File: subsys/net/ip/tcp.c [hunk 12]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+					K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
+			} else {
+				/* Acknowledge any FIN attempts, in case retransmission took
+				 * place.
+				 */
+				if ((FL(&fl, &, FIN, th_seq(th) + 1 == conn->ack)) || (len > 0)) {
+					tcp_out(conn, ACK);
+					verdict = NET_OK;
+
+					k_work_reschedule_for_queue(
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2862]: Is there a need to reset TIME_WAIT timer? It seems quite unsafe to me - a malfunctioning (or malicious) peer could keep sending FIN packets, effectively preventing us from releasing resources.

File: subsys/net/ip/tcp.c [hunk 13]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+		if (th) {
+			/*
+			 * Closing:
+			 * Our FIN has to be acknowledged
+			 * -   -> TCP_CLOSING
+			 * ACK -> TCP_TIME_WAIT
+			 */
+			int32_t new_len = tcp_compute_new_length(conn, th, len);
+			/* Count with -1 as the FIN flag has to be subtracted */
+			if ((new_len > -1) && (len > 0)) {
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2794]: Up for discussion, but how about extending `tcp_compute_new_length()` arguments with additional information about FIN (`bool fin_received`) so that the function could take this information into account? IMO it'd be cleaner, if we could always simply check whether `new_len > 0`. This check against `-1` is not very clear to me (although I do understand the rationale behind it).

**Review by rlubos — COMMENTED** [review]
Thanks, I think it's much easier to follow now with separate state handlers. I've left some suggestions.

File: subsys/net/ip/tcp.c [hunk 14]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-				/* Acknowledge any received data */
-				if (len > 0) {
 					tcp_out(conn, ACK);
-				}
-			} else if (th) {
-				NET_DBG("Staying in FIN_WAIT_1 flags 0x%x state seq %u, ack %u",
-					fl, conn->seq, conn->ack);
-				if (len > 0) {
+				} else {
+					/* Already acknowledged, we can go further */
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:2715]: Fixed

File: subsys/net/ip/tcp.c (hunk 13 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2794]: Good suggestion

File: subsys/net/ip/tcp.c [hunk 15]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+
+				tcp_out(conn, RST);
+				break;
+			}
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
+					"state seq %u, ack %u"
+					, conn->seq, conn->ack);
 				tcp_send_timer_cancel(conn);
 				next = TCP_FIN_WAIT_2;
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:2704]: You are right, it gives more direct insight for the reader what is going on.

File: subsys/net/ip/tcp.c [hunk 16]
```diff
@@ -2639,124 +2666,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
 				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
+			}
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check the sequence number of the FIN flag with tha ack
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Fixed

File: subsys/net/ip/tcp.c (hunk 11 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Done

**Comment by ssharks** [comment]
@rlubos Very thanks for the feedback. If you are deep in the matter, you quickly overlook that your code might not be very readable. Useful suggestions.

**Review by rlubos — APPROVED** [review]

File: subsys/net/ip/tcp.c [hunk 17]
```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-			K_MSEC(CONFIG_NET_TCP_TIME_WAIT_DELAY));
+			if (new_len > 0) {
+				/* This should not happen here, as no data can be send after
+				 * the FIN flag has been send.
+				 */
+				NET_ERR("conn: %p, new bytes %u during TIME-WAIT state "
+					"sending reset", conn, new_len);
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				tcp_out(conn, RST);
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2862]: And here.

File: subsys/net/ip/tcp.c [hunk 18]
```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+			} else {
+				/* Acknowledge any FIN attempts, in case retransmission took
+				 * place.
+				 */
+				if ((FL(&fl, &, FIN, th_seq(th) + 1 == conn->ack)) || (len > 0)) {
+					tcp_out(conn, ACK);
```
Comment (rlubos) [inline subsys/net/ip/tcp.c:2868]: And here.

**Review by rlubos — CHANGES_REQUESTED** [review]
I did some additional hands-on testing, when investigating another issue (Zephyr vs Zephyr) and encountered some issues, as stated in the comments - please take a look. 

Cherry-picking this PR was already a great improvement compared to what we have on `main`, but with suggested changes it worked flawlessly (the case was when both ends triggered FIN/ACK sequence independently).

**Comment by AndreyDodonov-EH** [comment]
> @AndreyDodonov-EH : Could you have a check at this revised PR?

I will have a look at the code, but I'm currently on vacations and don't have real hardware near me. @clamattia @kica-z could you give it a spin on the real HW?

File: subsys/net/ip/tcp.c [hunk 19]
```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
+				if (fin_acked) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:2838]: Fixed by moving the call out of the if statment

File: subsys/net/ip/tcp.c [hunk 20]
```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+				NET_DBG("FIN received, going to TIME WAIT");
+
+				next = tcp_enter_time_wait(conn);
+
 				verdict = NET_OK;
+				tcp_out(conn, ACK);
+			} else {
+				if (len > 0) {
+					/* Send out a duplicate ACK */
+					tcp_out(conn, ACK);
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c:2788]: To enter the TCP_FIN_WAIT_2 state tcp_send_timer_cancel, must have been called before, so I do not think it is required here.

File: subsys/net/ip/tcp.c (hunk 17 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2862]: To enter the TCP_TIME_WAIT state, tcp_send_timer_cancel, must have been called before, so I do not think it is required here. The only way to enter TIME_WAIT is through tcp_enter_time_wait, which calls tcp_send_timer_cancel. Packets with RST, ACK, PSH or ACK | PSH do not trigger the retransmit timer,

File: subsys/net/ip/tcp.c (hunk 18 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2868]: Same here

File: subsys/net/ip/tcp.c [hunk 21]
```diff
@@ -2639,124 +2671,205 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+				fin_acked = true;
+
+				verdict = NET_OK;
+			}
+
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check with the ack
+			 */
+			if ((FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) || (len > 0)) {
```
[the remark this answers is not in the document]
Comment (ssharks) [reply subsys/net/ip/tcp.c]: Good point, also switched to use net_tcp_seq_cmp at all locations where math is done with the sequence number to ensure correct overflowing behavior.

**Comment by ssharks** [comment]
> I did some additional hands-on testing, when investigating another issue (Zephyr vs Zephyr) and encountered some issues, as stated in the comments - please take a look.
> 
> Cherry-picking this PR was already a great improvement compared to what we have on `main`, but with suggested changes it worked flawlessly (the case was when both ends triggered FIN/ACK sequence independently).

Thanks for the thorough feedback and testing. Nice that you observed major improvement. I took over most of your suggestions.

File: subsys/net/ip/tcp.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (rlubos) [reply subsys/net/ip/tcp.c:2793]: Right, I've hit the issue in `TCP_CLOSING`, but that's probably not the case after we got an ACK for our FIN.

**Review by rlubos — DISMISSED** [review]
Thanks, re-tested with the latest revision and it works as expected now, the teardown goes smoothly w/o any unneeded retransmissions.

**Comment by ssharks** [comment]
Rebase to re-trigger CI

**Comment by ssharks** [comment]
Another rebase to trigger the CI. There has been a problem with the CI causing a lot of failures over the weekend.

File: subsys/net/ip/tcp.c [hunk 22]
```diff
@@ -2639,124 +2671,221 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-				}
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2699]: Could you add connection identifier here so that it is possible to debug things if there are multiple streams going on at the same time. So something like `conn %p` like in other debug prints in this file.

File: subsys/net/ip/tcp.c [hunk 23]
```diff
@@ -2639,124 +2671,221 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
-				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
 
-				/* Acknowledge any received data */
-				if (len > 0) {
+				/* State path is dependent on if the acknowledge is in */
+				if (fin_acked) {
+					/* Already acknowledged, we can go further */
+					NET_DBG("FIN received, going to TIME WAIT");
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2718]: same here

File: subsys/net/ip/tcp.c [hunk 24]
```diff
@@ -2639,124 +2671,221 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+			 */
+			if (FL(&fl, &, FIN, net_tcp_seq_cmp(th_seq(th) + len, conn->ack) == 0)) {
 				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				verdict = NET_OK;
-			} else if (len > 0) {
-				/* Send out a duplicate ACK */
-				tcp_out(conn,  ACK);
+				NET_DBG("FIN received, going to TIME WAIT");
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2784]: ditto

File: subsys/net/ip/tcp.c [hunk 25]
```diff
@@ -2639,124 +2671,221 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+				net_stats_update_tcp_seg_drop(conn->iface);
+
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to TIME WAIT "
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2826]: ditto

**Review by jukkar — COMMENTED** [review]
LGTM, minor comment about debug prints

File: subsys/net/ip/tcp.c (hunk 22 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2699]: Good suggestion

File: subsys/net/ip/tcp.c (hunk 23 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2718]: Done

File: subsys/net/ip/tcp.c (hunk 24 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2784]: Done

File: subsys/net/ip/tcp.c (hunk 25 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c:2826]: Done

**Review by jukkar — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by ssharks** [comment]
> > @AndreyDodonov-EH : Could you have a check at this revised PR?
> 
> I will have a look at the code, but I'm currently on vacations and don't have real hardware near me. @clamattia @kica-z could you give it a spin on the real HW?

Did you or your co-workers had the opportunity to validate if your original issue has been resolved?

**Comment by ssharks** [comment]
@AndreyDodonov-EH 
Thanks for testing. It is both good and bad news at the same time. Good to hear that the network behavior is as expected.

Spontaneous restarts could be multiple things, but on of them is a stack overflow. This happened to me on multiple occasions and these problems are hard to debug. This change puts in multi NET_ERR statements, of which some could be triggered in bad connectivity conditions. These logging statements do need a significant amount of stack to be handled correctly.

Could you check the CONFIG_NET_TCP_LOG_LEVEL you are using, you can print the stack usage by https://docs.zephyrproject.org/latest/services/debugging/thread-analyzer.html.
Or you could increase the stack size of the TCP work queue to see if that helps using CONFIG_NET_TCP_WORKQ_STACK_SIZE

Another option is to lower the log level of the newly introduced NET_ERR statements, by changing them to a value below your selected log level.

**Comment by AndreyDodonov-EH** [comment]
@ssharks 
Now that's a very good point about stack overflow, I will try it out tomorrow.
Truth be told, though, I'm still confused as to when Zephyr shows Hard Fault, and when just restarts silently.
Is it depending on what actually is in the memory outside the valid stack?

**Comment by ssharks** [comment]
I'm not so into the details on fault situations. I do know there are devices with hardware stack protection, but to what extend this is integrated into Zephyr and if it is default enabled I don't know.

**Comment by AndreyDodonov-EH** [comment]
@ssharks 
Isn't it CONFIG_NET_RX_STACK_SIZE that has to be increased?
`tcp_in` fin states are invoked (at least that's what debugger showed me once) in thread rx_q[0]

**Comment by AndreyDodonov-EH** [comment]
I wonder if it is possible to simulate that on QEMU with packet drops...

**Comment by ssharks** [comment]
@AndreyDodonov-EH It looks like there has been found a potential fix for the issue that you have observed see https://github.com/zephyrproject-rtos/zephyr/pull/63601
